Adding tests in CI

TonioGela opened the pull request:
After some brainstorming with @armanbilge, I added some tests to the CI to compile some scripts using scala-cli.

The whole logic is simple:
- `toolkitTesting / Test / test` depends on `toolkit.jvm / publishLocal` so that before running the tests the artifact gets published locally
- [BuildInfo](https://github.com/sbt/sbt-buildinfo) was added to have both `scalaversion` and `version` available in the test scope
- scala-cli was added as a dependency and it's called like `ScalaCli.main(Array("compile", "<tempFileName>"))`

Two specialized method were added, to test both scala 3.x.x and 2.13.x syntax (and toolkit artifact)

Solves https://github.com/typelevel/toolkit/issues/50.

TonioGela commented:
@armanbilge at the moment, the jvm tests are called even in the js and native jobs. I fear this might cause flakiness regarding the artifact that will get published, and if maybe this will solve

```scala
lazy val root = tlCrossRootProject
  .aggregate(toolkit, toolkitTest)
  .configureJVM(
    _.settings(
      (Test / test) := (Test / test)
        .dependsOn(toolkitTesting / Test / test)
        .value
    )
  )
```

WDYT?

armanbilge commented:
> at the moment, the jvm tests are called even in the js and native jobs

They are? I don't think so.

TonioGela commented:
> > at the moment, the jvm tests are called even in the js and native jobs
> 
> They are? I don't think so.

Look at the CI /shrug

armanbilge commented on build.sbt:
```diff
@@ -11,7 +11,11 @@ ThisBuild / mergifyStewardConfig ~= {
-lazy val root = tlCrossRootProject.aggregate(toolkit, toolkitTest)
+lazy val root = tlCrossRootProject
+  .aggregate(toolkit, toolkitTest)
+  .settings(
+    (Test / test) := (Test / test).dependsOn(toolkitTesting / Test / test).value
+  )
```
Oh, why did you add this? This is why its being called in the JS/Native jobs.

Instead, you should add `toolkitTesting` to the aggregate.

TonioGela replied:
hmm, you might be right

TonioGela replied:
it fixed!

TonioGela commented:
> Edit: so they are. but why 🤔

Well, it's `(Test/test)` that was redefined to depend on `toolkit.jvm / publishLocal`

TonioGela commented:
A thing to note is that `ScalaCli.main` calls `System.exit(1)` when it fails to compile, so the first broken test will fail fast for everything else. CI will still work and report its broken state, but we won't get a fancy output about the tests that are actually broken.

armanbilge commented:
Hmmm, that's actually quite annoying. What if we invoked scala-cli as an external process?

armanbilge commented on build.sbt:
```diff
@@ -44,6 +45,36 @@ lazy val toolkitTest = crossProject(JVMPlatform, JSPlatform, NativePlatform)
+  .in(file("toolkit-testing"))
+  .settings(
+    name := "toolkit-testing",
+    Test / test := (Test / test).dependsOn(toolkit.jvm / publishLocal).value,
```
It doesn't have to be this PR, but eventually I'd like to publish the JS and Native toolkits and test those as well.

Specifically, we seem to have an issue sometimes where the toolkit requires JS or Native versions newer than what the latest scala-cli is shipping by default, I wonder if we should try to avoid merging those ...
- https://github.com/typelevel/toolkit/issues/49

TonioGela replied:
Hmm, this can be done easily, but using scala-cli as an external process, a thing that will solve the `System.exit(1)` too.

That's another point in favor of this solution 🤔

TonioGela commented:
> Hmmm, that's actually quite annoying. What if we invoked scala-cli as an external process?

We could do that but I appreciate the leanness of the way is implemented now. I think another viable alternative might be to invoke the inner compile command internally and handle eventual `Left`, `None`, `Failure` or exceptions.

armanbilge commented:
> think another viable alternative might be to invoke the inner compile command internally

Ah that's the other thing: I don't want to specialize to `compile`, I think it would be good if we actually try to run the scripts. Otherwise we won't pick up on issues like https://github.com/typelevel/toolkit/pull/105#discussion_r1335073863.

I know it's not as lean, but I think an external process is the way to go long-term.

TonioGela commented:
> but I think an external process is the way to go long-term.

Hmm, let me think, what do we need apart calling the external process? Will adding `VirtusLab/scala-cli-setup@main` in the ci.yml be enough?

armanbilge commented:
We don't need to install anything, we already have scala-cli on the classpath since we added it as a dependency. We can load the classpath into the buildinfo and use that to invoke scala-cli as an ordinary java app.

TonioGela commented:
> Hmm, let me think, what do we need apart calling the external process? Will adding `VirtusLab/scala-cli-setup@main` in the ci.yml be enough?

The best way to know is to try, right? I'll rework this PR a bit tomorrow and let you know :)

armanbilge commented:
Here's how we do something similar in Cats Effect for running `IOApp` tests.

https://github.com/typelevel/cats-effect/blob/a6c5b155be3b729af49eb26204e101f2d527fa80/ioapp-tests/src/test/scala/IOAppSpec.scala#L58-L69

TonioGela commented:
> Here's how we do something similar in Cats Effect for running `IOApp` tests.
> 
> https://github.com/typelevel/cats-effect/blob/a6c5b155be3b729af49eb26204e101f2d527fa80/ioapp-tests/src/test/scala/IOAppSpec.scala#L58-L69

This definitively requires a night of sleep before trying to be understood. I'll take a look at it tomorrow, thanks ;)

TonioGela commented:
I like the idea; it will require a bit of trial and error, but it's feasible. I'll bring this PR back to Draft to make some tests. I have a question, though: why the full classpath was "injected" as a Java property while the platform was injected using BuildInfo?

TonioGela commented:
> I have a question, though: why the full classpath was "injected" as a Java property while the platform was injected using BuildInfo?

Because a setting cannot depend on a task ofc

TonioGela commented:
@armanbilge wdyt of the last commit? It's probably a bit overkilled, but it seems to work. Now it's the case to test both compilation and **running** for all the other platforms.

One thing I'm not 100% sure is how to structure the js and native tests: maybe it's the case to restore something like `Test / test := (Test / test).dependsOn(toolkit.jvm / publishLocal).value,` to let the test run after every platform has published locally its own artifact?

armanbilge commented on build.sbt:
```diff
@@ -44,6 +46,40 @@ lazy val toolkitTest = crossProject(JVMPlatform, JSPlatform, NativePlatform)
+        ExclusionRule("com.lihaoyi", "os-lib_2.13")
+      )
+    ),
+    Test / fork := true,
+    Test / javaOptions += s"-Dtoolkit.testing.classpath=${(Test / fullClasspath).value
+        .map(_.data.getAbsolutePath)
+        .mkString(File.pathSeparator)}"
```
Instead of this, we can just path the classpath as a `BuildInfo` key. Then we also don't need to hack.

Also `Compile / dependencyClasspath` should be sufficient if we can move scala-cli out of test scope.

TonioGela replied:
You could do this only if `Compile / dependencyClasspath` is a setting. You can't depend on tasks, neither using `.value` in settings definition. I'll try but I'm not sure.

armanbilge replied:
Ahhh. Well that's annoying, forgot that. Ok, so then maybe we need to pass it as an env variable so that JS and Native can pick it up.

TonioGela replied:
😎 
```scala
    buildInfoKeys +=
      BuildInfoKey.map(Compile / dependencyClasspath) { case (k, v) =>
        (k, v.seq.map(_.data.getAbsolutePath).mkString(File.pathSeparator))
      }
```

TonioGela replied:
This can be used in the CE build too :)

armanbilge commented on build.sbt:
```diff
@@ -44,6 +46,40 @@ lazy val toolkitTest = crossProject(JVMPlatform, JSPlatform, NativePlatform)
+  .in(file("toolkit-testing"))
+  .settings(
+    name := "toolkit-testing",
+    Test / test := (Test / test).dependsOn(toolkit.jvm / publishLocal).value,
```
let's make this a `crossProject(JVM, JS, Native)`. Then in `{jvm,js,native}Settings` we can do the appropriate:
```scala
Test / test := (Test / test).dependsOn(toolkit.{jvm,js,native}/ publishLocal).value
```

TonioGela replied:
hmm, and how can we pull the JVM dep of scala-cli? should we declare it w/o the `%%%` ?

armanbilge replied:
> should we declare it w/o the `%%%` ?

Yeah it's cheaty but would totally work. The other option is we put the dependency in its own project somewhere in the build instead of on this project, and just use that project as a way to generate a useful classpath.

TonioGela commented:
I think we're close. Probably scala js doesn't have `Process[F]` support, in that case :oldmanyellsatscalajs:

TonioGela commented:
@armanbilge we should be done :)

armanbilge commented on toolkit-testing/shared/src/test/scala/org/typelevel/toolkit/ScalaCliProcess.scala:
```diff
@@ -0,0 +1,101 @@
+          .through(Files[F].writeUtf8(path))
+          .compile
+          .drain
+          .as(path.toString)
+          .toResource
```
If you use `evalTap`.

```suggestion
```

armanbilge commented on toolkit-testing/shared/src/test/scala/org/typelevel/toolkit/ToolkitCompilationTest.scala:
```diff
@@ -0,0 +1,129 @@
+
+class ToolkitCompilationTest extends CatsEffectSuite {
+
+  testCompilation213("Toolkit should compile a simple Hello Cats Effect") {
```
Is it worthwhile to test compile and running separately? If so, then maybe we should do it without so much duplication.

armanbilge commented on toolkit-testing/shared/src/test/scala/org/typelevel/toolkit/ToolkitCompilationTest.scala:
```diff
@@ -0,0 +1,129 @@
+       |}"""
+  }
+
+  testRun213("Toolkit should run a script with every dependency") {
```
Maybe we can add one for `toolkit-test` as well.

armanbilge commented on toolkit-testing/shared/src/test/scala/org/typelevel/toolkit/ToolkitCompilationTest.scala:
```diff
@@ -0,0 +1,129 @@
+       |  def run = IO.println("Hello toolkit!")"""
+  }
+
+  def testCompilation213 = testCompilation("2.13.12")
+
+  def testCompilation3 = testCompilation("3.3.1")
+
+  def testRun213 = testCompilation("2.13.12")
+
+  def testRun3 = testCompilation("3.3.1")
```
Can we get these version numbers out of the `BuildInfo`? By loading the `crossScalaVersions` setting or something.

armanbilge commented on toolkit-testing/shared/src/test/scala/org/typelevel/toolkit/ToolkitCompilationTest.scala:
```diff
@@ -0,0 +1,129 @@
+       |}"""
+  }
+
+  testRun213("Toolkit should run a simple Hello Cats Effect") {
+    """|import cats.effect._
+       |
+       |object Hello extends IOApp.Simple {
+       |  def run = IO.println("Hello toolkit!")
+       |}"""
+  }
```
Another strategy we could use is:
```scala
  testRun("Toolkit should run a simple Hello Cats Effect") {
    if (scala3) // derived from BuildInfo
      ...
    else
      ...
  }
```

Then our CI logs don't need to include the ignored tests, which is a bit weird and noisy, and inconsistent with cross-platform (why don't we see ignored JS and Native tests too? but that would be crazy). It kind of defeats the purpose of the matrix.

armanbilge reviewed (commented):
This is really great! TIL a bunch of stuff too 🤓  Sorry, I just have way too many nitpicks 😅

TonioGela commented:
@armanbilge I've addressed most of the suggestions but:
- I can't get toolkit-test to work at all, even locally. Shouldn't `//> using test.toolkit typelevel:latest` be enough?
- Running on scala.js seems to fail to find the main class

armanbilge commented:
> * I can't get toolkit-test to work at all,

What's the error? Are you publishing it locally?

---

> * Running on scala.js seems to fail to find the main class

Sorry running what exactly? And what is the exact error?

TonioGela commented:
> What's the error? Are you publishing it locally?

Nope, I can't recall how to use the toolkit-test artifact, we haven't written a line on doc on how to use it 😓

armanbilge commented on build.sbt:
```diff
@@ -44,6 +46,62 @@ lazy val toolkitTest = crossProject(JVMPlatform, JSPlatform, NativePlatform)
+    buildInfoKeys += "scala3" -> (scalaVersion.value.head == '3')
+  )
+  .jvmSettings(
+    Test / test := (Test / test).dependsOn(toolkit.jvm / publishLocal).value,
```
What I meant is, we need to do `toolkitTest.jvm / publishLocal` as well :)

TonioGela replied:
Yep

armanbilge commented:
> Nope, I can't recall how to use the toolkit-test artifact,

You don't need to do anything, it's automatically in scope when you use `scala-cli test` command :)

TonioGela commented:
> ```
> java.lang.Exception: Non zero exit code (1) for run /tmp/3wmKco/.scala
> ```
> 
> Something seems wrong with the filename?

O_O Is the `Files[IO].tmpFile` impl that works differently in scala.js? Let me check.

armanbilge commented on tests/shared/src/test/scala/org/typelevel/toolkit/ScalaCliProcess.scala:
```diff
@@ -0,0 +1,86 @@
+      scriptBody: String
+  ): Resource[IO, String] =
+    Files[IO]
+      .tempFile(None, "", ".scala", None)
```
I think fs2.js has a problem if the prefix is blank, we should probably fix that in fs2.

TonioGela commented:
> > Nope, I can't recall how to use the toolkit-test artifact,
> 
> You don't need to do anything, it's automatically in scope when you use `scala-cli test` command :)

Using `scala-cli test` on this:
```
//> using toolkit typelevel::latest

import cats.effect.*
import munit.*

class Test extends CatsEffectSuite:
  test("test")(IO.unit)
```
or on the same but with `test.toolkit` always returns an error about munit not being found

armanbilge commented:
`typelevel::latest` it should be a single colon.

TonioGela commented:
**CI's green**

<img src="https://media4.giphy.com/media/IwAZ6dvvvaTtdI8SD5/giphy.gif"/>

TonioGela commented:
**CI's green**

<img src="https://media4.giphy.com/media/IwAZ6dvvvaTtdI8SD5/giphy.gif"/>

TonioGela commented:
There's something that I don't understand about the native pipeline, it takes more than 60 seconds to compile the first snippet 🤔

armanbilge reviewed (approved):
Very nicely done 👏

TonioGela commented:
> The real question is: how could have taken just 30 seconds to do all of this before?

Some of the CI runs that passed before mangling the general timeout took more than 30 seconds O_O'
Is 38 seconds still less than 30 seconds?

armanbilge commented:
The 30 second timeout applies per-test, not per-suite. So maybe that's how? or was a single test taking 38 seconds?

TonioGela commented:
> The 30 second timeout applies per-test, not per-suite. So maybe that's how? or was a single test taking 38 seconds?

Single test apparently: https://github.com/typelevel/toolkit/actions/runs/6309440614/job/17129364102

TonioGela commented:
Still, that's not a mistery that we'll solve here @armanbilge :D

Whenever you want,sir 

<img src="https://media3.giphy.com/media/cFkiFMDg3iFoI/giphy.gif"/>
